Ray_Serve/CV-Serve.ipynb

```diff
@@ -0,0 +1,1173 @@
+    "-   Architecture / Under-the-hood\n",
+    "    -   Ray cluster perspective - processes / workers / actors\n",
+    "    -   Request routing, queuing, load balancing in Serve\n",
+    "\n",
+    "### Example scenario: computer vision services\n",
+    "\n",
+    "For our example use case, we’ll see how to leverage Ray Serve to host a CV segmentation\n",
+    "model and how to enhance it using additional services such as image preprocessing.\n",
+    "\n",
+    "### Context: Ray AIR\n",
+    "\n",
+    "Ray AIR is the Ray AI Runtime, a set of high-level easy-to-use APIs for\n",
+    "ingesting data, training models – including reinforcement learning\n",
+    "models – tuning those models and then serving them.\n",
+    "\n",
+    "<img src=\"https://technical-training-assets.s3.us-west-2.amazonaws.com/Introduction_to_Ray_AIR/e2e_air.png\" width=600 loading=\"lazy\"/>\n",
+    "\n",
+    "Key principles behind Ray and Ray AIR are\n",
+    "* Performance\n",
+    "* Developer experience and simplicity"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "dfc41d8f-4332-4359-b957-f26bc79de7a3",
+   "metadata": {},
+   "source": [
+    "# Ray Serve\n",
+    "\n",
+    "Serve is a microservices framework for serving ML – the model serving\n",
+    "component of Ray AIR.\n",
+    "\n",
+    "<img src='https://technical-training-assets.s3.us-west-2.amazonaws.com/Ray_Serve/serve_architecture.png' width=700/>\n",
+    "\n",
+    "# Deployments\n",
+    "\n",
+    "`Deployment` is the fundamental user-facing element of serve.\n",
+    "\n",
+    "<img src='https://technical-training-assets.s3.us-west-2.amazonaws.com/Ray_Serve/deployment.png' width=600/>\n",
+    "\n",
+    "## Our First Service\n",
+    "\n",
+    "Let’s jump right in and get something simple up and running on Ray\n",
+    "Serve."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "48953d75-e69a-4fe3-adf8-079edc953683",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "@serve.deployment\n",
+    "class TextConverter:\n",
+    "    def convert(self, text):\n",
+    "        return \"***\" + str.upper(text) + \"***\""
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "62641bcc-f2a0-48ee-ac1d-182688552af4",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "app_handle = serve.run(TextConverter.bind())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "bcf7ce21-ee2a-48cb-8043-e986a4522e89",
+   "metadata": {},
+   "source": [
+    "## Key APIs and concepts\n",
+    "\n",
+    "`Deployment` represents a service and is created with the `@serve.deployment` decorator\n",
+    "* As end users, we don't instantiate `Deployment`s directly\n",
+    "* Ray will create them as actors, per our scaling requirements\n",
+    "\n",
+    "A __bound deployment__ is created with the `.bind` class method on the deployment class\n",
```

Comment:
A bound deployment is a serve deployment handle. The explanation here is a bit confusing. Maybe it's easier to say the .bind() method is used to bind the deployment with the arguments to the constructor, this returns a serve handle that can be passed to other deployments or used to start a service via serve run

Reply:
Is it fair to say that a ServeHandle and an Ingress Deployment are just special types of deployments?

Reply:
IngressDeployment is basically just the top level deployment in an application (the one you would pass to serve.run or in the import path in a yaml). Serve Handle is a "handle" to a deployment, that you can use to call into a deployment. So serve handle is used eg. when you have model 1 -> model 2

Reply:
For example, on the [Key Concepts](https://docs.ray.io/en/master/serve/key-concepts.html) page, the structure is currently:

1. Deployment
2. Application
3. ServeHandle (composing deployments)
4. Ingress Deployment (HTTP handling)
5. Deployment Graph

And this structure makes it seem like 3 and 4 are completely different ideas than 1, rather than an extension of 1. Is it accurate to reorder them like this:

1. Deployment
    a. Special types of deployment
            i. ServeHandle (composing deployments)
            ii. Ingress Deployment (HTTP handling)
3. Application
4. ~Deployment Graph~

Reply:
ServeHandle is not a type of deployment. IngressDeployment is. ServeHandle is a "pointer" to a deployment

Reply:
I agree this is confusing... trying to find a clear way to explain it.

The big sticking point for audiences is that a user can create a bound deployment any time, but not all bound deployments are serve handles. Some get automatically promoted/converted to serve handles at run time.

The docs say this: "When binding a deployment, you can include references to other bound deployments. Then, at runtime each of these arguments is converted to a [ServeHandle](https://docs.ray.io/en/latest/serve/api/doc/ray.serve.handle.RayServeHandle.html#ray.serve.handle.RayServeHandle) that can be used to query the deployment using a Python-native API."

And TBH that makes no sense to folks trying to learn the system. In my experience the average developer gives up trying to make sense of the logic and just relies on copy-pasting stuff that worked. I'm not personally saying I don't think the design is elegant, but the naming/explanation is a problem

Reply:
what do you mean by "but not all bound deployments are serve handles."? The return type when you bind is a deployment serve handle, and that can be passed around to other deployments or serve run

Reply:
So the technical bit is Deployment.bind() currently returns an Application (https://docs.ray.io/en/latest/_modules/ray/serve/deployment.html#Deployment.bind) which can be passed to serve.run, which in turn returns a serve handle. 

The human communication bit is that a developer has some situations where they can go foo = MyDeployment.bind() and then foo.bar.remote() and it "works" (typically inside another Deployment) but in other cases it does not (typically "outside" of a graph/application).

In the former case, the bound deployment has been promoted to a live serve handle when the graph that it's part of got .run(); in the latter case, no Application/graph was "run" ... but that distinction is hidden from the end user. In a strongly typed language (and dev community), there would be less of a problem -- some element of the type would almost certainly show a distinction, even if the purpose wasn't clear ... but in the Python world it's problematic.

Reply:
Maybe we can discuss this live. The return type of deployment.bind can be used to make remote calls to a deployment or in serve.run . I'm not sure what outside of an application means, this is how we should do it: use serve.run and make calls to ingress deployment over HTTP, ingress deployment can then call subsequent deployments using the previously initialized output of .bind() that are passed in to it.

Reply:
Since everyone is busy before Summit and there's the meeting detox, let me take another swing at it on my end when I revise the class -- including things like minimizing Graph API and adding FastAPI -- next week. I'll try and keep this part simpler and focus on the "happy path" (i.e.., proper patterns) that you described. If it still seems funky after that, then we can discuss live and decide how to further improve it.

Sound good? if not, I'm of course always open to meeting.

Reply:
yeah SGTM